src/Microsoft.Framework.DesignTimeHost/ProtocolManager.cs

```diff
@@ -0,0 +1,149 @@
+        /// </summary>
+        public const string EnvDthProtocol = "DTH_PROTOCOL";
+
+        /// <summary>
+        /// Type name of the message representing the protocol negotiation.
+        /// </summary>
+        public const string NegotiationMessageTypeName = "ProtocolVersion";
+
+        /// <summary>
+        /// Type name of the message representing the protocol negotiation.
+        /// </summary>
+        public const string InitializeMessageTypeName = "Initialize";
+
+        private readonly bool _environmentOverriden;
+
+        public ProtocolManager()
+        {
+            // initialized to the highest supported version or environment overriden value
+            int? protocol = GetProtocolVersionFromEnvironment();
+            
+            if (protocol.HasValue)
+            {
+                CurrentVersion = protocol.Value;
+                _environmentOverriden = true;
+            }
+            else
+            {
+                CurrentVersion = SupportedProtocolVersions[0];
+                _environmentOverriden = false;
+            }
+        }
+
+        public int CurrentVersion { get; private set; }
+
+        public bool NeedToRespond { get; private set; } = false;
+
+        public void Negotiate(Message message)
+        {
+            Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Initialize the protocol negotication.");
+
+            if (_environmentOverriden)
+            {
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: DTH protocol negotiation is override by environment variable {EnvDthProtocol} and set to {CurrentVersion}.");
+                return;
+            }
+
+            if (message.MessageType == NegotiationMessageTypeName)
+            {
+                if (message.ContextId != 0)
+                {
+                    return;
+                }
+
+                InternalNegotiate(message, needRespond: true);
+            }
+            else if (message.MessageType == InitializeMessageTypeName)
+            {
+                InternalNegotiate(message, needRespond: false);
+            }
+        }
+
+        private void InternalNegotiate(Message message, bool needRespond = false)
+        {
+            var tokenValue = message.Payload?["Version"];
+            if (tokenValue == null)
+            {
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication failed. Version property is missing in payload.");
+                return;
+            }
+
+            var preferredVersion = tokenValue.Value<int>();
+            if (preferredVersion == 0)
+            {
+                // the preferred version can't be zero. either property is missing or the the payload is corrupted.
+                Logger.TraceInformation($"[{nameof(ProtocolManager)}]: Protocol negotication failed. Protocol version 0 is invalid.");
+                return;
+            }
+
+            int? protocol = null;
```

Comment:
Might as well use `0` instead of `null` because you know from that the value cannot be `0`